src/App.tsx

```diff
@@ -1,29 +1,288 @@
       const startTimer = () => {
-         const myTimer = setTimeout(() => {
-            console.log(count, firstDimCount);
-            setCount(count => count + firstDimCount);
+         timerIdRef.current = setTimeout(() => {
+            timerExpiredCallback.current();
             startTimer();
-         }, 1000);
+         }, clockSpeedRef.current);
       };
-   
-    startTimer();
-      return () => {};
+      startTimer();
+
+      // if we ever unmount / destroy this component instance, clear the timeout
+      return () => clearTimeout(timerIdRef.current);
    }, []);
 
+   //-------------------------------------------
+   // FROM HERE
+
+   const handleTickBtnClick = () => {
+      clockSpeedRef.current = clockSpeedRef.current * (1 - 0.11);
+      setAntimatter(prevPrice => prevPrice - tickspeedPrice);
+      setTickspeedPrice(prevPrice => prevPrice * 10);
+   };
+
+   useEffect(() => {
+      canIstillBuyRef.current = antimatter - tickspeedPrice;
+   }, [tickspeedPrice]);
+
+   /* ref is needed to get the current state */
+   const canIstillBuy = () => {
+      if (canIstillBuyRef.current > 0) return true;
+      else return false;
+   };
+
+   /* It works but I do not know if there are better ways */
+   const handleBuyMaxClick = () => {
+      const repeatMax = () => {
+         timeIdRef.current = setTimeout(() => {
+            if (canIstillBuy()) {
+               console.log(canIstillBuy(), canIstillBuyRef.current, tickspeedPrice);
+               handleTickBtnClick();
+               repeatMax();
+            }
+         }, 20);
+
+         return () => clearTimeout(timeIdRef.current);
+      };
+      repeatMax();
+   };
+
+   // creates constantly an uptodate object which is then saved later
+   useEffect(() => {
+      dataRef.current = {
+         // ...dataRef.current,
+         antimatter,
+         tickspeedPrice,
+         firstDimCount,
+         firstDimFactor,
+         firstDimPrice,
+         firstDimFactorCount,
+         secondDimCount,
+         secondDimFactor,
+         secondDimPrice,
+         secondDimFactorCount,
+         thirdDimCount,
+         thirdDimFactor,
+         thirdDimPrice,
+         thirdDimFactorCount,
+         resetGameCounter,
+         highesDim,
+         timerIdRef,
+         idRef,
+         clockSpeedRef,
+      };
+      //console.log('dataRef object update cycle: ', dataRef.current ?? '');
+   });
+
+   // saves the created object every minuit to localStorage
+   useEffect(() => {
+      const startSave = () => {
+         timerIdRef.current = setTimeout(() => {
+            //localStorage.removeItem('dataRef');
+            localStorage.setItem('data', JSON.stringify(dataRef.current));
+            console.log('data save cycle: ', dataRef.current);
+            startSave();
+         }, 60000);
+      };
+      startSave();
+
+      // if we ever unmount / destroy this component instance, clear the timeout
+      return () => clearTimeout(timerIdRef.current);
+   }, []);
```

Comment:
Mmmm I understand exactly why you're putting all the values in a ref.  I can't think of a _great_ way to do this more cleanly.  One thing that comes to mind is you could pass all the values into the effect like this:
```
useEffect(() => {
  // lots of cool code
}, [gameState]); // <- note I'm assuming you've put all the values into one piece of state
```

And then write an effect that checks _if a certain amount of time has passed_.  Rather than setting a timeout.

For example call `Date.now()`.  That will give you the time in milliseconds (the number of Milliseconds since Jan. 1st, 1970, the "beginning of all time" as far as computers are concerned).  Save that value in `gameState.lastSaveTime`.

Whenever this effect hook is called (it will get called whenever gameState is changed, check if `Date.now() >= gameState.lastSaveTime + 60000`.  If it's greater, then save the game, and set `lastSaveTime` to the new `Date.now()` value.

This only works because our gameState is constantly changing due to the other timeout, and those changes can trigger our effect.  We're basically assuming (quite correctly, I might add), that if gameState has not changed, there's no need to save the game, regardless of how much time has passed.